### PR TITLE
Setting up mdBook for rust-bitcoin cookbook

### PR DESCRIPTION
**Pull request by Harshil-Jani**
Signed-off-by: Harshil Jani <harshiljani2002@gmail.com>

### CODE REVIEW

**Comment by Harshil-Jani** [comment]
@apoelstra @tcharding With this PR I have set up the mdBook template. Now, In past , I already had a [deployment workflow working in one of my mdBook](https://github.com/Harshil-Jani/GSOC-book/blob/main/.github/workflows/deploy.yml) . So, Can we use it here ? If yes then you'd need to create a `gh-pages` branch. 

Let me know about this and I can add another commit to put this deployment workflow here.

**Comment by apoelstra** [comment]
Do you know how I can preview this?

ACK in any case, thank you!

**Comment by apoelstra** [comment]
And your deployment script looks good to me, though please add a README here explaining what we need to do to make sure it triggers correctly :). I see it wants a `gh-pages` branch, but it actually triggers on changes to `main`, so it's not obvious to me what branch is what.

File: src/table_of_contents.md [hunk 1]
```diff
@@ -0,0 +1 @@
+# Table of Contents
```
Comment (Kixunil) [inline src/table_of_contents.md:1]: Is this auto-generated somehow?

**Review by Kixunil — APPROVED** [review]
ACK 66ebb587391ae97617e3e2cf39c11177d7a74dcc

Nice descriptions!

**Comment by Kixunil** [comment]
IIUC the script deploys the rendered book by committing it into gh-pages branch which should be somehow set up to show as the front page (how?). Am I correct?

File: src/table_of_contents.md (hunk 1 above)
Comment (Harshil-Jani) [reply src/table_of_contents.md:1]: No, I have added it in case if we need.

**Comment by Harshil-Jani** [comment]
> IIUC the script deploys the rendered book by committing it into gh-pages branch which should be somehow set up to show as the front page (how?). Am I correct?

Yes. mdBook when compiled from markdown renders the required `html` and `css` under `book/` directory. With this script, I am dumping entire `book/` into another branch i.e. `gh-pages` which separates from markdown contents. And after this all we need to do is go to github pages settings and select `gh-pages` as deployment branch. Since we already have every HTML we won't need any extra efforts. 
image

**Comment by Harshil-Jani** [comment]
> Do you know how I can preview this?
> 
> ACK in any case, thank you!

Locally you can always do 
1. Install mdBook
2. From the root directory run `mdbook serve --open`. 

To make it less of a burden, I have set up a preview version of this on my fork. 
https://harshil-jani.github.io/rust-bitcoin.github.io/

**Comment by apoelstra** [comment]
Thanks @Harshil-Jani, that worked great!

ACK 66ebb587391ae97617e3e2cf39c11177d7a74dcc

**Comment by tcharding** [comment]
Thanks @Harshil-Jani for working on this! I'm going to create a site for rust-bitcoin that includes the cookbook as part of it rather than the being the whole site. As such, can you do the following please then we can merge this
- leave index.html there
- put all the cookbook stuff in a directory called `cookbook`

For bonus points, if you can be bothered
- put a single newline on the end of all files (this is UNIX convention)
- use column width 100 (i.e., wrap lines at 100 characters), this is what we do for docs in the rest of rust-bitcoin repos

**Comment by tcharding** [comment]
No need to worry about the README build docs requested by @apoelstra, I can write all that when I add the full site.

**Comment by Kixunil** [comment]
Wow, it looks great!

**Pull request merged**